### PR TITLE
Bug 1883905: (fix) concurrent writes during image pulls

### PR DESCRIPTION
**Pull request by gallettilance**
**Description of the change:**

**Motivation for the change:**

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: pkg/lib/registry/registry.go [hunk 1]
```diff
@@ -157,35 +157,36 @@ func populate(ctx context.Context, loader registry.Load, graphLoader registry.Gr
...
-							if err != nil {
-								errs <- err
-							}
-							cleanups <- cleanup
-							overwriteImageMap[img.Bundle.Package][to] = from
+							unpacked <- unpackedImage{to: to, from: from, cleanup: cleanup, err: err}
 						} else {
-							overwriteImageMap[img.Bundle.Package][to] = from
+							unpacked <- unpackedImage{to: to, from: from, cleanup: func() { return }, err: nil}
+							// safe because this case is only hit once
```
Comment (ecordell) [inline pkg/lib/registry/registry.go:175]: it's only hit once per goroutine, but it's still concurrent access to a map

Comment (gallettilance) [reply pkg/lib/registry/registry.go:175]: what I meant is there is only one go routine that will ever hit this case

File: pkg/lib/registry/registry.go [hunk 2]
```diff
@@ -157,35 +157,37 @@ func populate(ctx context.Context, loader registry.Load, graphLoader registry.Gr
...
 					// parallelize image pulls
 					go func(bundle registry.BundleKey, img *registry.ImageInput) {
 						if bundle.CsvName != img.Bundle.Name {
 							to, from, cleanup, err := unpackImage(ctx, reg, image.SimpleReference(bundle.BundlePath))
-							if err != nil {
-								errs <- err
-							}
-							cleanups <- cleanup
-							overwriteImageMap[img.Bundle.Package][to] = from
+							unpacked <- unpackedImage{to: to, from: from, cleanup: cleanup, err: err}
```
Comment (ecordell) [inline pkg/lib/registry/registry.go:172]: if `unpackImage` were updated to return an `unpackedImage`, this whole goroutine could just be:

```go
go func(bundle registry.BundleKey, img *registry.ImageInput) {
  unpacked <- unpackImage(ctx, reg, image.SimpleReference(bundle.BundlePath))
}
```

Comment (gallettilance) [reply pkg/lib/registry/registry.go:172]: I think we want to skip the one we already unpacked

File: pkg/lib/registry/registry.go [hunk 3]
```diff
@@ -157,35 +157,37 @@ func populate(ctx context.Context, loader registry.Load, graphLoader registry.Gr
...
+				for i := 0; i < len(bundles); i++ {
+					unpack := <-unpacked
+					if unpack.err != nil {
+						return unpack.err
+					}
+					overwriteImageMap[img.Bundle.Package][unpack.to] = unpack.from
+					if _, ok := unpackedImageMap[unpack.to]; ok {
+						delete(unpackedImageMap, unpack.to)
 					}
+					defer unpack.cleanup()
```
Comment (ankitathomas) [inline pkg/lib/registry/registry.go:190]: Shouldn't we cleanup bundles if they have errors?

Comment (gallettilance) [reply pkg/lib/registry/registry.go:190]: ideally yes but it depends on the error if there is anything to clean up

**Pull request merged**